README.md
## Ogólny opis :smiley:

Front na podstawie bootstrap template sb admin2. Historia wersji z opisami w CHANGELOG'u

Dodana klasa DB2 - realizująca połączenie z bazą danych oraz CRUD
Dodana klasa User - rejestracja,logowanie użytkownika
Dodana klasa PageController - przekierowania/ przeładowania strony

skrypt logout.php - służący do wylogowywania - ale niestety nie resetuje sesji - do poprawy


Widoki
- login.php
- register.php
- index.php

to są w zasadzie widokami html z wcięciami echo php.

## SQL - Struktura tabel 
```sql
CREATE TABLE `Users` (
  `Id` varchar(13) NOT NULL,
  `Login` varchar(333) NOT NULL,
  `Pass` varchar(333) NOT NULL,
  `Name` varchar(333) NOT NULL,
  `Surname` varchar(333) NOT NULL,
  `Gender` varchar(33) NOT NULL,
  `DateCreated` date NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;

ALTER TABLE `Users`
  ADD PRIMARY KEY (`Id`);
COMMIT;
```

## Połącznie z bazą danych
Konfiguracja w pliku database.php
Zmieniamy
```php
    private $servername = "localhost";
    private $username = "dbuser1";
    private $password = "1234";
    private $myDB   = "wskz";
    
```


## Thanks :pray:

CHANGELOG.md
##
    0.2.1 - klasy php
##
    0.2.0 - struktury sql
## 
    0.1.2 - klasa database.php
##
    0.1.1 - start repo
## 
    0.1.0 - usunięcie z szablonu 'sb admin2' niepotrzebnych elementów

index.php
<?php 
  require 'php/PageController.php';
  session_start();
  $Page = new PageController();
  if($_SESSION['logged']==false)
    $Page->RedirectIfNotLogged();
    
?>

<!DOCTYPE html>
<html lang="en">

<head>

  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="">
  <meta name="author" content="">

  <title>SB Admin 2 - Blank</title>

  <!-- Custom fonts for this template-->
  <link href="vendor/fontawesome-free/css/all.min.css" rel="stylesheet" type="text/css">
  <link href="https://fonts.googleapis.com/css?family=Nunito:200,200i,300,300i,400,400i,600,600i,700,700i,800,800i,900,900i" rel="stylesheet">

  <!-- Custom styles for this template-->
  <link href="css/sb-admin-2.min.css" rel="stylesheet">

</head>

<body id="page-top">

  <!-- Page Wrapper -->
  <div id="wrapper">

    <!-- Sidebar -->
    <ul class="navbar-nav bg-gradient-primary sidebar sidebar-dark accordion" id="accordionSidebar">

      <!-- Sidebar - Brand -->
      <a class="sidebar-brand d-flex align-items-center justify-content-center" href="index.html">
        <div class="sidebar-brand-icon rotate-n-15">
          <i class="fas fa-laugh-wink"></i>
        </div>
        <div class="sidebar-brand-text mx-3">SB Admin <sup>2</sup></div>
      </a>

      <!-- Divider -->
      <hr class="sidebar-divider my-0">

      <!-- Nav Item - Dashboard -->
      <li class="nav-item">
        <a class="nav-link" href="index.html">
          <i class="fas fa-fw fa-tachometer-alt"></i>
          <span>Dashboard</span></a>
      </li>

      <!-- Divider -->
      <hr class="sidebar-divider">


      <!-- Sidebar Toggler (Sidebar) -->
      <div class="text-center d-none d-md-inline">
        <button class="rounded-circle border-0" id="sidebarToggle"></button>
      </div>

    </ul>
    <!-- End of Sidebar -->

    <!-- Content Wrapper -->
    <div id="content-wrapper" class="d-flex flex-column">

      <!-- Main Content -->
      <div id="content">

        <!-- Topbar -->
        <nav class="navbar navbar-expand navbar-light bg-white topbar mb-4 static-top shadow">

          <!-- Sidebar Toggle (Topbar) -->
          <button id="sidebarToggleTop" class="btn btn-link d-md-none rounded-circle mr-3">
            <i class="fa fa-bars"></i>
          </button>

          <!-- Topbar Navbar -->
          <ul class="navbar-nav ml-auto">

            <!-- Nav Item - Search Dropdown (Visible Only XS) -->
            <li class="nav-item dropdown no-arrow d-sm-none">
              <a class="nav-link dropdown-toggle" href="#" id="searchDropdown" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                <i class="fas fa-search fa-fw"></i>
              </a>
              <!-- Dropdown - Messages -->
              <div class="dropdown-menu dropdown-menu-right p-3 shadow animated--grow-in" aria-labelledby="searchDropdown">
                <form class="form-inline mr-auto w-100 navbar-search">
                  <div class="input-group">
                    <input type="text" class="form-control bg-light border-0 small" placeholder="Search for..." aria-label="Search" aria-describedby="basic-addon2">
                    <div class="input-group-append">
                      <button class="btn btn-primary" type="button">
                        <i class="fas fa-search fa-sm"></i>
                      </button>
                    </div>
                  </div>
                </form>
              </div>
            </li>

            <div class="topbar-divider d-none d-sm-block"></div>

            <!-- Nav Item - User Information -->
            <li class="nav-item dropdown no-arrow">
              <a class="nav-link dropdown-toggle" href="#" id="userDropdown" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                <span class="mr-2 d-none d-lg-inline text-gray-600 small"><?= $_SESSION['User']['Name'].' '.$_SESSION['User']['Surname']?></span>
                <img class="img-profile rounded-circle" src="https://source.unsplash.com/QAB-WJcbgJk/60x60">
              </a>
              <!-- Dropdown - User Information -->
              <div class="dropdown-menu dropdown-menu-right shadow animated--grow-in" aria-labelledby="userDropdown">
                <a class="dropdown-item" href="#">
                  <i class="fas fa-user fa-sm fa-fw mr-2 text-gray-400"></i>
                  Profile
                </a>
                <a class="dropdown-item" href="#">
                  <i class="fas fa-cogs fa-sm fa-fw mr-2 text-gray-400"></i>
                  Settings
                </a>
                <a class="dropdown-item" href="#">
                  <i class="fas fa-list fa-sm fa-fw mr-2 text-gray-400"></i>
                  Activity Log
                </a>
                <div class="dropdown-divider"></div>
                <a class="dropdown-item" href="php/User.php?logout=1" data-toggle="modal" data-target="#logoutModal">
                  <i class="fas fa-sign-out-alt fa-sm fa-fw mr-2 text-gray-400"></i>
                  Logout
                </a>
              </div>
            </li>

          </ul>

        </nav>
        <!-- End of Topbar -->

        <!-- Begin Page Content -->
        <div class="container-fluid">

          <!-- Page Heading -->
          <h1 class="h3 mb-4 text-gray-800">Witaj <?=$_SESSION['User']['Name'].' '.$_SESSION['User']['Surname']?></h1>
          <h4><?=' ('.$_SESSION['User']['Gender'].')'?>

        </div>
        <!-- /.container-fluid -->

      </div>
      <!-- End of Main Content -->

      <!-- Footer -->
      <footer class="sticky-footer bg-white">
        <div class="container my-auto">
          <div class="copyright text-center my-auto">
            <span>Copyright &copy; Your Website 2020</span>
          </div>
        </div>
      </footer>
      <!-- End of Footer -->

    </div>
    <!-- End of Content Wrapper -->

  </div>
  <!-- End of Page Wrapper -->

  <!-- Scroll to Top Button-->
  <a class="scroll-to-top rounded" href="#page-top">
    <i class="fas fa-angle-up"></i>
  </a>

  <!-- Logout Modal-->
  <div class="modal fade" id="logoutModal" tabindex="-1" role="dialog" aria-labelledby="exampleModalLabel" aria-hidden="true">
    <div class="modal-dialog" role="document">
      <div class="modal-content">
        <div class="modal-header">
          <h5 class="modal-title" id="exampleModalLabel">Ready to Leave?</h5>
          <button class="close" type="button" data-dismiss="modal" aria-label="Close">
            <span aria-hidden="true">×</span>
          </button>
        </div>
        <div class="modal-body">Select "Logout" below if you are ready to end your current session.</div>
        <div class="modal-footer">
          <button class="btn btn-secondary" type="button" data-dismiss="modal">Cancel</button>
          <a class="btn btn-primary" href="login.php">Logout</a>
        </div>
      </div>
    </div>
  </div>

  <!-- Bootstrap core JavaScript-->
  <script src="vendor/jquery/jquery.min.js"></script>
  <script src="vendor/bootstrap/js/bootstrap.bundle.min.js"></script>

  <!-- Core plugin JavaScript-->
  <script src="vendor/jquery-easing/jquery.easing.min.js"></script>

  <!-- Custom scripts for all pages-->
  <script src="js/sb-admin-2.min.js"></script>

</body>

</html>


login.php

<!DOCTYPE html>
<html lang="en">

<head>

  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="">
  <meta name="author" content="">

  <title>SB Admin 2 - Login</title>

  <!-- Custom fonts for this template-->
  <link href="vendor/fontawesome-free/css/all.min.css" rel="stylesheet" type="text/css">
  <link href="https://fonts.googleapis.com/css?family=Nunito:200,200i,300,300i,400,400i,600,600i,700,700i,800,800i,900,900i" rel="stylesheet">

  <!-- Custom styles for this template-->
  <link href="css/sb-admin-2.min.css" rel="stylesheet">

</head>

<body class="bg-gradient-primary">

  <div class="container">

    <!-- Outer Row -->
    <div class="row justify-content-center">

      <div class="col-xl-10 col-lg-12 col-md-9">

        <div class="card o-hidden border-0 shadow-lg my-5">
          <div class="card-body p-0">
            <!-- Nested Row within Card Body -->
            <div class="row">
              <div class="col-lg-6 d-none d-lg-block bg-login-image"></div>
              <div class="col-lg-6">
                <div class="p-5">
                  <div class="text-center">
                    
                    <?php if(isset($_GET['error'])):?>
                      <h4 class="h6 text-danger mb-4 mt-2">Niepoprawny login lub hasło</h4>
                    <?php else:?>
                      <h1 class="h4 text-red-900 mb-4">Welcome Back!</h1>
                    <?php endif;?>
                  </div>
                  <form class="user" action="php/User.php" method="POST">
                    <div class="form-group">
                      <input type="email" class="form-control form-control-user" name="Login" aria-describedby="emailHelp" placeholder="Login...">
                    </div>
                    <div class="form-group">
                      <input type="password" class="form-control form-control-user" name="Password" placeholder="Password">
                    </div>
                    
                    <button type="Submit" name="loginUser" value="1" class="btn btn-primary btn-user btn-block">Zaloguj</button>
                    
                  </form>
                  <hr>
                 
                  <div class="text-center">
                    <a class="small" href="register.php">Utwórz konto!</a>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>

      </div>

    </div>

  </div>

  <!-- Bootstrap core JavaScript-->
  <script src="vendor/jquery/jquery.min.js"></script>
  <script src="vendor/bootstrap/js/bootstrap.bundle.min.js"></script>

  <!-- Core plugin JavaScript-->
  <script src="vendor/jquery-easing/jquery.easing.min.js"></script>

  <!-- Custom scripts for all pages-->
  <script src="js/sb-admin-2.min.js"></script>

</body>

</html>


php/User.php
<?php
require 'Database.php';
require 'PageController.php';

class User{
    private $db;
    private $page;
    private $salt = 'salt12345';
    // private $login;
    // private $password;
    // private $name;
    // private $surname;
    // private $gender;

    public function __construct(){
        $this->db = new DB2();
        $this->page = new PageController();
    }

    public function Hash($pass){
        return md5($pass.$this->salt);
    }

    public function findUser($name,$pass=null){
        if($pass==null)
            $where = "Login='".$name."'";
        else
            $where = "Login='".$name."' AND Password='".$this->Hash($pass)."'";
    
       return $this->db->GetUser('User',$where);    
    }

    public function Login($postArray){
        if($postArray['Login'] && $postArray['Password']){

            $this->user=$this->findUser($postArray['Login'],$postArray['Password']);    
            
            if($this->user){
                               
                $_SESSION['logged']=true;

                $_SESSION['User']= $this->user;

                $this->page->Redirect('../index.php');
            }else{
                $this->page->Redirect('../login.php?error=1');
            }
        }else
            return false;
    }

    public function New($postArray){
        
        if($postArray['Login'] && $postArray['Password']){
            
            $this->user = $this->findUser($postArray['Login']);
            if($this->user)
                $this->page->Redirect('../register.php?error=2');
            else{
                unset($postArray['registerNewUser']);

                $postArray['Password'] = $this->Hash($postArray['Password']);
                $result = $this->db->Insert('User',$postArray);
                
                if($result){
                    $this->page->Redirect('../login.php');
                }
            }
        }
    }

}

session_start();

$user = new User();

if(!empty($_POST['registerNewUser'])){
    $user->New($_POST);
}
if(!empty($_POST['loginUser'])){
    $user->Login($_POST);
}


//------session logout 
//- not working dont know why?
if(isset($_GET['logout']) && $_GET['logout']==1)
{
    // remove all session variables
    session_unset();
    // destroy the session
    session_destroy();
}


php/database.php
<?php
class DB2{
    private $conn;  //connection
    private $servername = "localhost";
    private $username = "dbuser1";
    private $password = "1234";
    private $myDB   = "wskz";

    public function __construct(){
        $this->Connect();
    }
    
    public function Connect(){
        try {
            $this->conn = new PDO("mysql:host=$this->servername;dbname=$this->myDB", $this->username, $this->password);
            $this->conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
        }
        catch(PDOException $e) {
            echo "Error: " . $e->getMessage();
        }
    }
    public function GetUser($table_name,$where=null){
        $sql="SELECT * FROM $table_name ";
        if($where)
            $sql.="WHERE $where";
    
        $stmt = $this->conn->prepare($sql);

        $stmt->execute(); 
        $result = $stmt->fetch();
        
        return($result);
    }
    public function Insert($table_name, $values){
        $this->conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
        $values['DateCreated']=date('Y-m-d');
        $values['Id']=uniqid();
    
        $sql = "INSERT INTO $table_name (".implode(',',array_keys($values)).") VALUES (:";
        $sql.=implode(',:',array_keys($values)).")";
    
        $stmt = $this->conn->prepare($sql);
        $ar=[];
    
        foreach($values as $key=>$v){
                
            $stmt->bindParam(":$key", $ar[$key]);
            
            $ar[$key]=$v; 
        }
        $stmt->execute();
        return true;
    }
    public function Update($table_name, $id, $values){
        $this->conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
        $values['DateModified']=date('Y-m-d');
        $vals='';
        $i=0;
        foreach($values as $key=>$v){       
            if($i!=0)
                $vals .=",";
    
            $vals .=" $key='".$v."'";
            $i++; 
        }
        $sql = "UPDATE $table_name SET $vals WHERE Id='$id'";

        $stmt = $this->conn->prepare($sql);
        $stmt->execute();
    }
    public function Delete($table_name, $id){
        $this->conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
    
        $sql = "DELETE FROM $table_name WHERE Id='";
        
        if(is_array($id))
            $sql.=implode("' OR Id='",$id)."'";    
        else
            $sql.=$id."'";
    
        $stmt = $this->conn->prepare($sql);
        $stmt->execute();
    }

}

register.php
<!DOCTYPE html>
<html lang="en">

<head>

  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="">
  <meta name="author" content="">

  <title>SB Admin 2 - Register</title>

  <!-- Custom fonts for this template-->
  <link href="vendor/fontawesome-free/css/all.min.css" rel="stylesheet" type="text/css">
  <link href="https://fonts.googleapis.com/css?family=Nunito:200,200i,300,300i,400,400i,600,600i,700,700i,800,800i,900,900i" rel="stylesheet">

  <!-- Custom styles for this template-->
  <link href="css/sb-admin-2.min.css" rel="stylesheet">

</head>

<body class="bg-gradient-primary">

  <div class="container">
    
    <div class="card o-hidden border-0 shadow-lg my-5">
      <div class="card-body p-0">
        <!-- Nested Row within Card Body -->
        <div class="row">
          <div class="col-lg-5 d-none d-lg-block bg-register-image"></div>
          <div class="col-lg-7">
            <div class="p-5">
              <div class="text-center">
                
                <?php if(isset($_GET['error'])):?>
                      <h4 class="h6 text-danger mb-4 mt-2">Taki login już istnieje !!!</h4>
                <?php else:?>
                      <h1 class="h4 text-gray-900 mb-4">Utwórz nowe konto</h1>
                <?php endif;?>
              </div>
              <form class="user" action="php/User.php" method="POST">
                <div class="form-group-row mb-2">
                  <center>
                    <div class="form-check form-check-inline">
                      <input class="form-check-input" type="radio" name="Gender" id="inlineRadio1" value="Mężczyzna" required>
                      <label class="form-check-label" for="inlineRadio1">Mężczyzna</label>
                    </div>
                    <div class="form-check form-check-inline">
                      <input class="form-check-input" type="radio" name="Gender" id="inlineRadio2" value="Kobieta">
                      <label class="form-check-label" for="inlineRadio2">Kobieta</label>
                    </div>
                  </center>
                </div>
                <div class="form-group row">
                  <div class="col-sm-6 mb-3 mb-sm-0">
                    <input type="text" class="form-control form-control-user" name="Name" placeholder="Imie" required>
                  </div>
                  <div class="col-sm-6">
                    <input type="text" class="form-control form-control-user" name="Surname" placeholder="Nazwisko" required>
                  </div>
                </div>
                
                <div class="form-group">
                  <input type="email" class="form-control form-control-user" name="Login" placeholder="Email Address" required>
                </div>
                <div class="form-group row">
                  <div class="col-sm-6 mb-3 mb-sm-0">
                    <input type="password" class="form-control form-control-user" name="Password" pattern=".{8,}"   required title="8 znaków minimum" placeholder="Hasło" >
                  </div>
                  <div class="col-sm-6">
                    <input type="password" class="form-control form-control-user" id="exampleRepeatPassword" placeholder="Powtórz hasło" required>
                  </div>
                </div>
                <button type="Submit" name="registerNewUser" value="1" class="btn btn-primary btn-user btn-block">Zarejestruj</button>
              </form>
              <hr>
              <div class="text-center">
                <a class="small" href="login.php">Masz już konto? Zaloguj się!</a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>

  </div>

  <!-- Bootstrap core JavaScript-->
  <script src="vendor/jquery/jquery.min.js"></script>
  <script src="vendor/bootstrap/js/bootstrap.bundle.min.js"></script>

  <!-- Core plugin JavaScript-->
  <script src="vendor/jquery-easing/jquery.easing.min.js"></script>

  <!-- Custom scripts for all pages-->
  <script src="js/sb-admin-2.min.js"></script>

</body>

</html>
